integrations/tensorflow/compiler/LegalizeTF.cpp

```diff
@@ -0,0 +1,124 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/Shape/IR/Shape.h"      // from @llvm-project
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // from @llvm-project
+#include "mlir/Pass/Pass.h"                   // from @llvm-project
+#include "mlir/Support/LLVM.h"                // from @llvm-project
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/chlo_ops.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/rewriters.h"
+#include "tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.h"
+#include "tensorflow/compiler/mlir/xla/transforms/passes.h"
+
+namespace mlir {
+namespace mhlo {
+namespace {
+
+class LegalizeTF : public PassWrapper<LegalizeTF, FunctionPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<chlo::HloClientDialect, mhlo::MhloDialect,
+                    shape::ShapeDialect, StandardOpsDialect>();
+  }
+
+ public:
+  LegalizeTF() = default;
+  LegalizeTF(const LegalizeTF &) {}
+
+  /// Performs the lowering to XLA dialect.
+  void runOnFunction() override {
+    auto op = getFunction();
+    MLIRContext *context = op.getContext();
+    OwningRewritePatternList canonicalizePatterns;
+    for (auto *op : context->getRegisteredOperations())
+      op->getCanonicalizationPatterns(canonicalizePatterns, context);
+
+    OwningRewritePatternList patterns;
+    // Note that the `OperationConverter` orders patterns lexicographically by:
+    // 1) Ascending legalization depth (i.e., minimum number of patterns
+    // necessary
+    //    to arrive at conversion target).
+    // 2) Descending pattern benefit.
+    // 3) Order of patterns in `OwningRewritePatternList`.
+
+    // Add TF->HLO legalization patterns.
+    PopulateLegalizeTfPatterns(context, &patterns);
+
+    // Add TF->TF lowering patterns.
+    TF::PopulateLoweringTFPatterns(context, &patterns);
+
+    // Populate with CHLO->HLO lowerings to account for TF ops legalized to
+    // CHLO first.
+    chlo::PopulateLegalizeChloToHloPatterns(context, &patterns);
+
+    // ConstantLike op is convenient to create splat constants, but is
+    // canonicalized to plain HLO constant if statically shaped. Add the
+    // canonicalization pattern to pattern list to enable multi-hop lowering.
+    chlo::ConstantLikeOp::getCanonicalizationPatterns(patterns, context);
+
+    ConversionTarget target(*context);
+    target.addIllegalDialect<chlo::HloClientDialect>();
+    target.addLegalDialect<MhloDialect>();
+    target.addLegalDialect<StandardOpsDialect>();
+    target.addLegalDialect<shape::ShapeDialect>();
+    target.addLegalOp<CallOp>();
+    target.addLegalOp<TensorCastOp>();
+
+    DenseSet<Operation *> prevUnconvertedOps;
+    DenseSet<Operation *> unconvertedOps;
+
+    while (true) {
```

Comment:
Please add a test for the fixed-point iteration logic here.

Reply:
The linspace test should test the iteration logic. I am not sure what other operations may run into this issue.

Reply:
I mean a .mlir test that exercises it.